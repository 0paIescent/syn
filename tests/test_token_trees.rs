extern crate syn;
use syn::TokenTree::{self, Token};
use syn::DelimToken::*;
use syn::Token::*;

#[test]
fn test_struct() {
    let raw = "
        #[derive(Debug, Clone)]
        pub struct Item {
            pub ident: Ident,
            pub attrs: Vec<Attribute>,
        }
    ";

    let expected =
        vec![Token(Pound),
             delimited(Bracket,
                       vec![ident("derive"),
                            delimited(Paren, vec![ident("Debug"), Token(Comma), ident("Clone")])]),
             ident("pub"),
             ident("struct"),
             ident("Item"),
             delimited(Brace,
                       vec![ident("pub"),
                            ident("ident"),
                            Token(Colon),
                            ident("Ident"),
                            Token(Comma),

                            ident("pub"),
                            ident("attrs"),
                            Token(Colon),
                            ident("Vec"),
                            Token(Lt),
                            ident("Attribute"),
                            Token(Gt),
                            Token(Comma)])];

    let result = syn::parse_token_trees(raw.parse().unwrap()).unwrap();
    if result != expected {
        panic!("{:#?}\n!=\n{:#?}", result, expected);
    }
}

fn delimited(delim: syn::DelimToken, tts: Vec<TokenTree>) -> TokenTree {
    TokenTree::Delimited(syn::Delimited {
                             delim: delim,
                             tts: tts,
                         })
}

fn ident(s: &str) -> TokenTree {
    TokenTree::Token(Ident(syn::Ident::new(s)))
}
