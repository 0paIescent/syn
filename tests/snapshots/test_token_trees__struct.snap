---
created: "2019-04-06T05:46:23.184384152Z"
creator: insta@0.7.4
source: tests/test_token_trees.rs
expression: syntax_tree
---
TokenStream [
    Punct {
        op: '#',
        spacing: Alone,
    },
    Group {
        delimiter: Bracket,
        stream: TokenStream [
            Ident {
                sym: derive,
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [
                    Ident {
                        sym: Debug,
                    },
                    Punct {
                        op: ',',
                        spacing: Alone,
                    },
                    Ident {
                        sym: Clone,
                    },
                ],
            },
        ],
    },
    Ident {
        sym: pub,
    },
    Ident {
        sym: struct,
    },
    Ident {
        sym: Item,
    },
    Group {
        delimiter: Brace,
        stream: TokenStream [
            Ident {
                sym: pub,
            },
            Ident {
                sym: ident,
            },
            Punct {
                op: ':',
                spacing: Alone,
            },
            Ident {
                sym: Ident,
            },
            Punct {
                op: ',',
                spacing: Alone,
            },
            Ident {
                sym: pub,
            },
            Ident {
                sym: attrs,
            },
            Punct {
                op: ':',
                spacing: Alone,
            },
            Ident {
                sym: Vec,
            },
            Punct {
                op: '<',
                spacing: Alone,
            },
            Ident {
                sym: Attribute,
            },
            Punct {
                op: '>',
                spacing: Joint,
            },
            Punct {
                op: ',',
                spacing: Alone,
            },
        ],
    },
]
